File: query_primer/query_primer__word_count.cs
using System.Linq;
using static System.Console;

internal class Program
{
    static void Main()
    {
        int[] n = ReadLine().Split().Select( int.Parse ).ToArray();
        int[] a = new int[n[0] + 1];

        SetOut( new System.IO.StreamWriter( OpenStandardOutput() ) { AutoFlush = false } );
        foreach( int i in Enumerable.Range( 1, n[0] ) ) {
            a[i] = a[i - 1] + int.Parse( ReadLine() );
        }
        foreach( int i in Enumerable.Range( 0, n[1] ) ) {
            int[] lr = ReadLine().Split().Select( int.Parse ).ToArray();
            int pa = lr[1] - lr[0] + 1;
            int pb = lr[3] - lr[2] + 1;

            if( n[0] / 3m <= pa ) {
                WriteLine( n[0] / 3m <= pb ? "DRAW" : "B" );
                continue;
            } else if( n[0] / 3m <= pb ) {
                WriteLine( 'A' );
                continue;
            }

            int ia = a[lr[1]] - a[lr[0] - 1];
            int ib = a[lr[3]] - a[lr[2] - 1];

            WriteLine( ib < ia ? "A" : ia < ib ? "B" : "DRAW" );
        }
        Out.Flush();
        return;
    }
}

File: query_primer/query_primer__word_count.py
n, k = list(map(int, input().split()))
a = [0]

for i in range(n):
    a.append(a[-1] + int(input()))
for _ in range(k):
    al, ar, bl, br = list(map(int, input().split()))
    pa = ar - al + 1
    pb = br - bl + 1

    if n / 3 <= pa:
        print("DRAW" if n / 3 <= pb else "B")
        continue
    elif n / 3 <= pb:
        print("A")
        continue

    ia = a[ar] - a[al - 1]
    ib = a[br] - a[bl - 1]

    print("A" if ib < ia else "B" if ia < ib else "DRAW")
# 実行時間 290ms
